src/lib/GUID.cpp

```diff
@@ -70,20 +122,72 @@ bool GUID::is_equal(GUID other)
 }
 
 // This is used for FileDataStores. The GUID is given as a string in the file
-void GUID::from_string(std::string str)
+void GUID::from_string(std::string const str)
 {
   ONE_DEBUG_MSG(("\n"));
 
-  Data1 = strtol(str.substr(0, 8).c_str(), NULL, 16);
-  Data2 = strtol(str.substr(9, 4).c_str(), NULL, 16);
-  Data3 = strtol(str.substr(14, 4).c_str(), NULL, 16);
-  Data4[0] = strtol(str.substr(19, 4).c_str(), NULL, 16);
-  Data4[1] = strtol(str.substr(24, 4).c_str(), NULL, 16);
-  Data4[2] = strtol(str.substr(28, 4).c_str(), NULL, 16);
-  Data4[3] = strtol(str.substr(32, 4).c_str(), NULL, 16);
+  if (str.size() < MinStringGUIDLength)
+  {
+    return;
+  }
+
+  if (str.front() == '{' && str.size() > MaxStringGUIDLength)
+  {
+    return;
+  }
+
+
+  size_t i{0};
+  str.at(i) == '{' ? i++ : 0;
+
+  Data1 = strtol(str.substr(i,8).c_str(), NULL, 16);
+  i += 8;
+
+  str.at(i) == '-' ? i++ : 0;
+
+  Data2 = strtol(str.substr(i,4).c_str(), NULL, 16);
+  i += 4;
+
+  str.at(i) == '-' ? i++ : 0;
+
+  Data3 = strtol(str.substr(i,4).c_str(), NULL, 16);
+  i += 4;
+
+  str.at(i) == '-' ? i++ : 0;
+
+  Data4[0] = strtol(str.substr(i,4).c_str(), NULL, 16);
+  i += 4;
+
+  str.at(i) == '-' ? i++ : 0;
+
+  Data4[1] = strtol(str.substr(i,4).c_str(), NULL, 16);
+  i += 4;
+
+  Data4[2] = strtol(str.substr(i,4).c_str(), NULL, 16);
+  i += 4;
+
+  Data4[3] = strtol(str.substr(i,4).c_str(), NULL, 16);
+  i += 4;
 
   ONE_DEBUG_MSG((" from string, dat good?\n"));
 
-  (void) str;
 }
+
+
+librevenge::RVNGInputStream *operator>>(librevenge::RVNGInputStream *input, GUID &obj)
```

Comment:
I'm having trouble wrapping my head around this, could you show me an example on how to use this?

Reply:
This piece overloads the `>>` operator. Defining it as a friend function and outside the class allows to write in a stream like fashion. Have a look at the example [`Header.cpp`](https://github.com/blu-base/libone/blob/guid_operator_overload/src/lib/Header.cpp). It makes parsing the guid class a tiny bit less verbose.

It's not really necessary. I wanted to see, whether you are open to this style of operator overloading.

Reply:
This actually seems really nice! I haven't done any operator overloading before and initially did this codebase coming more from a "pure C" point of view, so I'm all for using syntactic sugar whenever it makes sense.